src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/fsm/haflow/validation/HaFlowValidationService.java

```diff
@@ -0,0 +1,242 @@
+     persistence layer and converts it into a simple switch rule format. It then compares the expected speaker data with
+     the actual speaker data and creates a validation response object containing discrepancies if any.
+     @param flowId the ID of the flow to validate
+     @param flowDumpResponses the list of flow dump responses received from switches
+     @param metersDumpResponses the list of meter dump responses received from switches
+     @param groupDumpResponses the list of group dump responses received from switches
+     @param switchIdSet the set of switches for which speaker data is required
+     @return a HaFlowValidationResponse object containing validation results for the specified flow
+     @throws FlowNotFoundException if the specified flow is not found in the persistence layer
+     @throws SwitchNotFoundException if speaker data is not found for one or more switches specified in the switchIdSet
+     */
+    public HaFlowValidationResponse validateFlow(String flowId, List<FlowDumpResponse> flowDumpResponses,
+                                                 List<MeterDumpResponse> metersDumpResponses,
+                                                 List<GroupDumpResponse> groupDumpResponses,
+                                                 Set<SwitchId> switchIdSet)
+            throws FlowNotFoundException, SwitchNotFoundException {
+
+        Optional<HaFlow> foundFlow = haFlowRepository.findById(flowId);
+        if (!foundFlow.isPresent()) {
+            throw new FlowNotFoundException(flowId);
+        }
+        HaFlow flow = foundFlow.get();
+
+        if (flow.getForwardPath() == null) {
+            throw new InvalidPathException(flowId, "The forward path has not been returned.");
+        }
+        if (flow.getReversePath() == null) {
+            throw new InvalidPathException(flowId, "The reverse path has not been returned.");
+        }
+
+        List<SpeakerData> actualSpeakerData = flowDumpResponses.stream()
+                .flatMap(e -> e.getFlowSpeakerData().stream()).collect(Collectors.toList());
+        int rulesCount = actualSpeakerData.size();
+        actualSpeakerData.addAll(metersDumpResponses.stream()
+                .flatMap(e -> e.getMeterSpeakerData().stream()).collect(Collectors.toList()));
+        int metersCount = new Long(metersDumpResponses.stream()
+                .mapToLong(e -> e.getMeterSpeakerData().size()).sum()).intValue();
+        actualSpeakerData.addAll(groupDumpResponses.stream()
+                .flatMap(e -> e.getGroupSpeakerData().stream()).collect(Collectors.toList()));
+
+        Map<SwitchId, List<SimpleSwitchRule>> actualSimpleSwitchRulesBySwitchId
+                = simpleSwitchRuleConverter.convertSpeakerDataToSimpleSwitchRulesAndGroupBySwitchId(actualSpeakerData);
+
+
+        PersistenceDataAdapter dataAdapter = getDataAdapter(switchIdSet, flow.getPaths()
+                .stream().flatMap(haFlowPath1 -> haFlowPath1.getSubPaths().stream())
```

Comment:
haFlowPath1 is a strange name. Can we use haFlowPath?

Reply:
fixed